storage_proxy: do not fence reads and writes to local tables

patjed41 opened the pull request:
Fencing is necessary only for reads and writes to non-local tables.
Moreover, fencing a read or write to a local table can cause an
error on the bootstrapping node. It is explained in the comment
in storage_proxy::get_fence.

A scenario described in the comment has been reported in
scylladb/scylladb#16423. A write to the local RAFT table failed
because of fencing, and it killed server_impl::io_fiber.

Fixes scylladb/scylladb#16423

scylladb-promoter commented:
#### :green_circle: CI State: SUCCESS

:white_check_mark: - [Build](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/5380/)
:white_check_mark: - [Unit Tests](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/5380/)
:white_check_mark: - [dtest](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release/6792/)
#### Build Details:
- Duration: 2 hr 35 min
- Builder: spider2.cloudius-systems.com

patjed41 commented:
v2:
- added a specific scenario to the comment
- changed the commit message slightly - now it specifies that the errors this commit is fixing are possible only on the bootstrapping nodes

gusev-p commented on service/storage_proxy.cc:
```diff
@@ -1550,9 +1550,25 @@ class abstract_write_response_handler : public seastar::enable_shared_from_this<
+        auto& keyspace = _proxy->local_db().find_keyspace(_mutation_holder->schema()->ks_name());
+        auto replication_strategy_type = keyspace.get_replication_strategy().get_type();
+        auto fence = replication_strategy_type == locator::replication_strategy_type::local ?
+                fencing_token{} : fencing_token{_effective_replication_map_ptr->get_token_metadata().get_version()};
```
we can use `_effective_replication_map_ptr->get_replication_strategy()->get_type()`

And probably it's better to extract `get_fence` (e.g. as `storage_proxy::get_fence(erm)`), reuse it in both bases and move the comment inside it.

avikivity replied:
> we can use `_effective_replication_map_ptr->get_replication_strategy()->get_type()`


Probably this suggestion prevents a measurable regression. It's also bad form to use an unfenced lookup (find_keyspace) in a fencing call.

patjed41 replied:
> we can use `_effective_replication_map_ptr->get_replication_strategy()->get_type()`
> 
> And probably it's better to extract `get_fence` (e.g. as `storage_proxy::get_fence(erm)`), reuse it in both bases and move the comment inside it.

I applied both suggestions.

scylladb-promoter commented:
#### :green_circle: CI State: SUCCESS

:white_check_mark: - [Build](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/5385/)
:white_check_mark: - [Unit Tests](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/5385/)
:white_check_mark: - [dtest](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release/6800/)
#### Build Details:
- Duration: 3 hr 2 min
- Builder: spider7.cloudius-systems.com

patjed41 commented:
v3:
- used `_effective_replication_map_ptr->get_replication_strategy()` instead of loading the keyspace and getting a replication strategy from it
- moved `get_fence` to `storage_proxy` so that it is used for both reads and writes
- moved the comment with a specific scenario to `storage_proxy::get_fence`
- used `storage_proxy::get_fence` also in `abstract_write_response_handler::apply_remotely` -- the code is more clear and consistent this way

scylladb-promoter commented:
#### :green_circle: CI State: SUCCESS

:white_check_mark: - [Build](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/5457/)
:white_check_mark: - [Unit Tests](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/5457/)
:white_check_mark: - [dtest](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release/6877/)
#### Build Details:
- Duration: 1 hr 38 min
- Builder: spider8.cloudius-systems.com
